File: Programmars/Level_2/올바른 괄호.py
def solution(s):
    stack = []
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(s[i])
        else:
            if not stack:
                return False
            stack.pop()
    if stack:
        return False
    return True


print(solution("()()"))  # true
print(solution("(())()"))  # true
print(solution(")()("))  # false
print(solution("(()("))  # false

File: JavaScript/Programmars/Level_2/올바른 괄호.js
// 문제: https://programmers.co.kr/learn/courses/30/lessons/12909?language=javascript
const solution = (s) => {
  if (s.length % 2 !== 0) return false;

  const stack = [];
  for (let p of s) {
    if (p === '(') stack.push(p);
    else {
      if (stack.length < 0) return false;
      stack.pop();
    }
  }

  return stack.length === 0;
};

console.log(solution('()()')); // true
console.log(solution('(())()')); // true
console.log(solution(')()(')); // false
console.log(solution('(()(')); // false
